File: tests/code/corpus/langs/typescript/shapes.ts
import { PI } from "./constants";

export interface Drawable {
  draw(): void;
}

export type Radius = number;

export enum Kind {
  Circle,
  Square,
}

export class Shape implements Drawable {
  draw(): void {}
  area(): number {
    return 0;
  }
}

export class Circle extends Shape {
  constructor(private radius: Radius) {
    super();
  }
  area(): number {
    return PI * this.radius * this.radius;
  }
}

export function makeCircle(radius: Radius): Circle {
  return new Circle(radius);
}

function totalArea(shapes: Shape[]): number {
  return shapes.reduce((sum, shape) => sum + shape.area(), 0);
}

export { totalArea };

File: tests/code/corpus/langs/rust/shapes.rs
use std::f64::consts::PI;

pub struct Circle {
    pub radius: f64,
}

pub enum Kind {
    Round,
    Square,
}

pub trait Drawable {
    fn draw(&self);
}

impl Circle {
    pub fn area(&self) -> f64 {
        PI * self.radius * self.radius
    }

    pub fn describe(&self) -> String {
        format!("circle of area {}", self.area())
    }
}

pub fn make_circle(radius: f64) -> Circle {
    Circle { radius }
}

fn total_area(shapes: &[Circle]) -> f64 {
    shapes.iter().map(|shape| shape.area()).sum()
}
